src/permissions.rs

```diff
@@ -0,0 +1,63 @@
+use users::{ Groups, Users };
+use users::os::unix::GroupExt;
+
+pub fn check_group_membership<U: Users+ Groups>(cache: &U, expected_group_names: &[&str]) {
+    let mut is_one_group_found = false;
+    let mut i = 0;
+    let expected_group_size = expected_group_names.len();
+
+    while !is_one_group_found && i < expected_group_size {
+        let group_name = expected_group_names[i];
+
+        if let Some(group) = cache.get_group_by_name(group_name) {
+            is_one_group_found = true;
+            let current_uid = cache.get_current_uid();
+            let user = cache.get_user_by_uid(current_uid).unwrap();
+
+            if !group.members().contains(&user.name().to_owned()) {
+                panic!("Not a member of the {} group", group_name);
+            }
+        }
```

Comment:
I don't like this. This is checking whether the current user is in the group, not whether the current process is in the group. So this test can pass (user is a member of the group) when the process is not. For example, if you edited the /etc/group file to add the user, then this would pass even though the shell which did the editing isn't yet a member.

libc::getgroups returns the groups that this process is currently a member of. So I think we should go back to using that and then check each of those groups to see if its name matches one of the ones passed in.

uid shouldn't enter into the equation anywhere (and if it did we should be using geteuid and not getuid).

This is a subtle distinction, but one that I think is quite important.